cpp/src/Ice/Reference.h

```diff
@@ -238,48 +242,49 @@ class RoutableReference : public Reference
-    virtual ReferencePtr changeEndpointSelection(Ice::EndpointSelectionType) const;
-    virtual ReferencePtr changeLocatorCacheTimeout(int) const;
-
-    virtual ReferencePtr changeTimeout(int) const;
-    virtual ReferencePtr changeConnectionId(const std::string&) const;
-    virtual ReferencePtr changeConnection(const Ice::ConnectionIPtr&) const;
-
-    virtual bool isIndirect() const;
-    virtual bool isWellKnown() const;
-
-    virtual void streamWrite(Ice::OutputStream*) const;
-    virtual std::string toString() const;
-    virtual Ice::PropertyDict toProperty(const std::string&) const;
-
-    virtual bool operator==(const Reference&) const;
-    virtual bool operator<(const Reference&) const;
-
-    virtual ReferencePtr clone() const;
-
-    virtual RequestHandlerPtr getRequestHandler() const;
-    virtual BatchRequestQueuePtr getBatchRequestQueue() const;
+    std::vector<EndpointIPtr> getEndpoints() const final;
+    std::string getAdapterId() const final;
+    LocatorInfoPtr getLocatorInfo() const final;
+    RouterInfoPtr getRouterInfo() const final;
+    bool getCollocationOptimized() const final;
+    bool getCacheConnection() const final;
+    bool getPreferSecure() const final;
+    Ice::EndpointSelectionType getEndpointSelection() const final;
+    int getLocatorCacheTimeout() const final;
+    std::string getConnectionId() const final;
+    std::optional<int> getTimeout() const final;
+
+    ReferencePtr changeEncoding(const Ice::EncodingVersion&) const final;
+    ReferencePtr changeCompress(bool) const final;
+    ReferencePtr changeMode(Mode) const final;
+    ReferencePtr changeEndpoints(const std::vector<EndpointIPtr>&) const final;
+    ReferencePtr changeAdapterId(const std::string&) const final;
+    ReferencePtr changeLocator(const std::optional<Ice::LocatorPrx>&) const final;
+    ReferencePtr changeRouter(const std::optional<Ice::RouterPrx>&) const final;
+    ReferencePtr changeCollocationOptimized(bool) const final;
+    ReferencePtr changeCacheConnection(bool) const final;
+    ReferencePtr changePreferSecure(bool) const final;
+    ReferencePtr changeEndpointSelection(Ice::EndpointSelectionType) const final;
+    ReferencePtr changeLocatorCacheTimeout(int) const final;
+
+    ReferencePtr changeTimeout(int) const final;
+    ReferencePtr changeConnectionId(const std::string&) const final;
+    ReferencePtr changeConnection(const Ice::ConnectionIPtr&) const final;
+
+    bool isIndirect() const final;
+    bool isWellKnown() const final;
+
+    void streamWrite(Ice::OutputStream*) const final;
+    std::string toString() const final;
+    Ice::PropertyDict toProperty(const std::string&) const final;
+
+    bool operator==(const Reference&) const final;
+    bool operator<(const Reference&) const final;
+
+    ReferencePtr clone() const final;
+
+    RequestHandlerPtr getRequestHandler() const final;
+    const BatchRequestQueuePtr& getBatchRequestQueue() const final;
```

Comment:
For a fixed reference, getBatchRequestQueue() returns a non-null queue even when mode != batch.